src/Microsoft.OData.Edm/ExtensionMethods/ExtensionMethods.cs

```diff
@@ -2937,6 +2997,60 @@ internal static string ReplaceAlias(this IEdmModel model, string name)
             return name;
         }
 
+        #region IEdmTargetPath extensions
+
+        /// <summary>
+        /// Get the path string from the <see cref="IEdmTargetPath"/>.
+        /// </summary>
+        /// <param name="targetPath">The target path.</param>
+        /// <returns>The target path string.</returns>
+        public static string GetPathString(this IEdmTargetPath targetPath)
+        {
+            EdmUtil.CheckArgumentNull(targetPath, nameof(targetPath));
+
+            IEdmEntityContainer container = targetPath.Segments[0] as IEdmEntityContainer;
+
+            if (container == null)
+            {
+                throw new InvalidOperationException(Strings.TargetPath_FirstSegmentMustBeIEdmEntityContainer);
+            }
+
+            IEdmEntityContainerElement containerElement = targetPath.Segments[1] as IEdmEntityContainerElement;
+
+            if (containerElement == null)
+            {
+                throw new InvalidOperationException(Strings.TargetPath_SecondSegmentMustBeIEdmEntityContainerElement);
+            }
+
+            StringBuilder stringBuilder = new StringBuilder();
+            stringBuilder.Append(container.FullName());
+            stringBuilder.Append("/");
+            stringBuilder.Append(containerElement.Name);
+
+            int index = 2;
+
+            while (index < targetPath.Segments.Count)

```

Comment:
I wonder whether this could be optimized further since we can calculate the total number of segments in the path string ahead of time.

For example, we could create a fixed-size array of strings, add the segments and separators to that array, then call `string.Join` at the end. This array could be allocated on the stack (using `stackalloc`) or from array pool.

Alternatively, we could pre-compute the total number of characters in the path string (by summing up the `Segment.Name.Length` properties and the separators), then allocate a fixed-size array of chars (can be allocated using `stackalloc` if it's small enough or from array pool if profitable), then names of each segment into that array at the correct positions, then call `new String(char[])` at the end.

I don't know if these would be more performant than the `StringBuilder()` since it has its own internal optimizations, but it also does a lot and allocates temporary arrays as well. So, if we know the number of strings or total size of the combined strings ahead of time, it may be a good idea to explore other options.

Also, I don't know if this method is used in hot paths and whether it's worth the effort to try and optimize it.

Reply:
Good thing this method is called once per `EdmTargetPath`. But if target paths are created often, it could add up.

Reply:
Out of curiosity, I have created a simplified version of this problem to test the different implementations and found that creating a fixed sized array and using join yielded better overall performance, the versions with array-pooling and char array where slower than string.Join without array pooling (to my surprise), but used slightly less memory

| Method                   | pathName    | Mean      | Error      | StdDev    | Gen0   | Allocated |
|------------------------- |------------ |----------:|-----------:|----------:|-------:|----------:|
| WithStringBuilder        | longPath    | 274.86 ns | 126.345 ns |  6.925 ns | 0.2856 |    1232 B |
| WithJoin                 | longPath    | 138.62 ns |  29.393 ns |  1.611 ns | 0.1056 |     456 B |
| WithJoinAndArrayPool     | longPath    | 178.99 ns |  18.947 ns |  1.039 ns | 0.0834 |     360 B |
| WithCharArray            | longPath    | 198.61 ns |  32.978 ns |  1.808 ns | 0.1669 |     720 B |
| WithNoTempAllocCharArray | longPath    | 169.25 ns |  25.699 ns |  1.409 ns | 0.0834 |     360 B |
| WithStringBuilder        | shortPath   |  58.52 ns | 146.921 ns |  8.053 ns | 0.0352 |     152 B |
| WithJoin                 | shortPath   |  55.31 ns |  34.871 ns |  1.911 ns | 0.0222 |      96 B |
| WithJoinAndArrayPool     | shortPath   |  80.61 ns |  28.241 ns |  1.548 ns | 0.0111 |      48 B |
| WithCharArray            | shortPath   |  69.61 ns |  12.681 ns |  0.695 ns | 0.0241 |     104 B |
| WithNoTempAllocCharArray | shortPath   |  59.23 ns |   3.982 ns |  0.218 ns | 0.0111 |      48 B |
| WithStringBuilder        | verLongPath | 565.84 ns |  20.285 ns |  1.112 ns | 0.5779 |    2496 B |
| WithJoin                 | verLongPath | 412.97 ns | 214.625 ns | 11.764 ns | 0.2966 |    1280 B |
| WithJoinAndArrayPool     | verLongPath | 447.89 ns |  49.392 ns |  2.707 ns | 0.2408 |    1040 B |
| WithCharArray            | verLongPath | 586.33 ns | 327.525 ns | 17.953 ns | 0.4816 |    2080 B |
| WithNoTempAllocCharArray | verLongPath | 489.28 ns | 108.696 ns |  5.958 ns | 0.2403 |    1040 B |

You can find the benchmarks that I used here: https://gist.github.com/habbes/54f9fc9c28683e13d87db479fd829281

Again, bear in mind that this is a simplified version that does not account for the segment type casts (eg `is SchemaType` and the calls to `FullyQualifiedName()`

Reply:
I have updated the code to use `string.Join(...)`